packages/studio/server/src/studio/testing/recorder.ts

```diff
@@ -0,0 +1,83 @@
+import * as sdk from '@botpress/sdk'
+import _ from 'lodash'
+
+import { Scenario } from './typings'
+import { convertLastMessages } from './utils'
+
+export class Recorder {
+  private _lastEvent?: sdk.IO.IncomingEvent
+  private _scenario?: Scenario
+  private _target?: string
+
+  constructor(private logger: sdk.Logger) {}
+
+  async processIncoming(event: sdk.IO.IncomingEvent): Promise<void> {
+    if (!this.isRecording() || this._scenario?.initialState || !this._target) {
+      return
+    }
+
+    // TODO: This should most likely be a call to the runtime
+    const target = ''
+    //const target = await getMappingFromVisitor(this.database, this.logger, event.botId, this._target)
+    if (target === event.target) {
+      this._scenario!.initialState = event.state
+    }
+  }
+
+  async processCompleted(event: sdk.IO.IncomingEvent): Promise<void> {
+    if (!this.isRecording() || !this._target) {
+      return
+    }
+
+    // TODO: This should most likely be a call to the runtime
+    const target = ''
```

Comment:
or messaging ?

Reply:
Right! In any case, I think we should take some time to properly rethink the testing module considering that we now have the messaging server and the runtime as standalone. Maybe we can refactor and remove a lot of logic there.